File: 247_strobogrammatic_number_ii.rb
=begin
A strobogrammatic number is a number that looks the same when rotated 180 degrees (looked at upside down).

Find all strobogrammatic numbers that are of length = n.

Example:

Input:  n = 2
Output: ["11","69","88","96"]
=end

# @param {Integer} n
# @return {String[]}
def find_strobogrammatic(n)
    helper(n, n)
end

def helper(len, n)
    if len == 0
        return [""]
    end
    if len == 1
        return ["0", "1", "8"]
    end
    
    list = helper(len-2, n)
    ans = []
    (0...list.length).each do |i|
        if len != n
            ans << '0' + list[i] + '0'
        end
        ans << '1' + list[i] + '1'
        ans << '6' + list[i] + '9'
        ans << '8' + list[i] + '8'
        ans << '9' + list[i] + '6'
    end
    
    return ans
end

File: 247_strobogrammatic_number_ii.py
'''
A strobogrammatic number is a number that looks the same when rotated 180 degrees (looked at upside down).

Find all strobogrammatic numbers that are of length = n.

Example:

Input:  n = 2
Output: ["11","69","88","96"]
'''

class Solution:
    def findStrobogrammatic(self, n: int) -> List[str]:
        return self.helper(n, n)

    def helper(self, length, n):
        if length == 0:
            return [""]
        if length == 1:
            return ["0", "1", "8"]
        
        nums = self.helper(length - 2, n)
        ans = []
        for i in range(len(nums)):
            if length != n:
                ans.append(f"0{nums[i]}0")
            ans.append(f"1{nums[i]}1")
            ans.append(f"6{nums[i]}9")
            ans.append(f"8{nums[i]}8")
            ans.append(f"9{nums[i]}6")
        
        return ans

